### DISCUSSION TITLE
`RangedWeapon` reload always skips up to 2 frames, causing 33% slower Chaingun rate of fire (and up to 70% RoF/RPM reduction in extreme cases)

### DISCUSSION THREAD

**Discussion by Krzeszny** [FakeFishGames/Barotrauma#9483] [Code questions and modding support]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

**Disclaimer:** I've edited this report based on Regalis's comments and new testing ideas. Previously it was tested with a stopwatch and the results had 1 mistake (chaingun was 33% slower instead of 11% slower) and 1 typo (0.25 coilgun reload instead of 0.15)**

In the editor, I recorded videos in 60 fps showing how many frames it takes between shots of the default Coilgun, default Chaingun, Pulse Laser, Chaingun modified to 1 shot per second and 60 shots per second. Using the theoretical reloads (0.25, 0.1, modded 1 and modded 0.01667) and knowing that the game engine runs at 60 Hz, I calculated their actual reloads.

- The **Coilgun** skips 2 frames per reload, resulting in 13% slower firing than expected
- The **Chaingun** skips 2 frames per reload, resulting in **33%** slower firing than expected
- The **Laser** skips 2 frames per second, resulting in 2.6% slower firing than expected
- Chaingun mod. 60 skips 2 frames per shot when firing at 60 RPM, resulting in 3.3% slower firing than expected
- Chaingun mod. 3600 skips **1** frame per shot when firing at 3600 RPM, resulting in 100% slower firing than expected (theoretical maximum = 1800 RPM)

It has been like that at least since update 0.14.6.

![brave_ldezrQoa6w](https://user-images.githubusercontent.com/10529134/176167169-700cead6-10b2-417a-9286-6000d78b118c.png)

I've created a formula used to calculate the RoF. It works between a 0.01 second reload and infinity:
EU:
`=60/(1/60*IF(Reload=0,05;4;FLOOR(Reload/(1/60);1)+2))`
US:
`=60/(1/60*IF(Reload=0.05,4,FLOOR(Reload/(1/60),1)+2))`

### Reproduction steps

1. Load the Orca 2
2. Set its supercapacitors to 100%
3. Record firing a few shots at 60 fps
6. Calculate frames per shot in a video editing software
7. Compare the actual value with the expected value

### Bug prevalence

Happens every time I play

### Version

Oldest: 0.14.6 or older (Coilgun and small arms only, before the submarine gun update)
Newest: 0.18.12

### Which operating system did you encounter this bug on?

Windows

### Additional info

The highest possible RPM is currently 1800 (instead of 3600 or 6000 as expected from reload times), which is  1 shot per 2 frames.

**Comment by Regalis11** [comment]
I'm pretty sure this happens because 0.15 s isn't divisible by the duration of one frame (1/60 s). Reloading the coilgun would take a little over 9 frames, but since there's no such thing as a partial frame, it can only fire every 10 frames (every 0.16... s, which is a little over 11% more than the theoretical reload time). The error is of course larger the smaller the reload time is: if you had a weapon with a reload time of 0.000016s, it would take a thousand times longer in reality because the turret can't fire until the next frame (after 1/60 s).

In my opinion, this isn't something we really need to fix. The firing rates are only "wrong" if you calculate the theoretical firing rates "wrong", without taking into account how Euler integration works.

**Comment by Krzeszny** [comment]
I was playing at 100 fps so I'm assuming physics are capped to 60.

Anyway, it's fine if it's balanced, although it causes confusion when trying to calculate rounds per minute.

**Comment by Regalis11** [comment]
Yeah, by frames I mean the updates (physics, item/entity updates and such) which run at a fixed 60 Hz.

**Comment by Krzeszny** [comment]
> Reloading the coilgun would take a little over 9 frames, but since there's no such thing as a partial frame, it can only fire every 10 frames (every 0.16... s, which is a little over 11% more than the theoretical reload time).

Coilgun has a 0.25 s reload, not 0.15 s. 

Actually, 0.25 (coilgun) is dividable by 1/60 and so is 0.1 (chaingun). But it's even more complicated than that.
I've retested it by counting the frames in 60 fps video recordings (my game was running at 100 fps) in an open-source video editing program kdenlive.

1. The coilgun appears to shoot once every 17 frames instead of once every 15 frames, giving a 1.13(3)% longer reload. The box is emptied in 56.65 seconds or 3399 frames, also giving a 1.13(3)% longer reload. At least it's consistent, but it shouldn't be skipping 1 frame, let alone 2.
2. The chaingun appears to shoot once every 8 frames instead of once every 6 frames, which would give a 33.(3)% increase in reload. The box is emptied in 66.8 seconds or 4008 frames, giving a **33.(6)%** increase.
3. To test it further, I changed the chaingun's reload to 0.01, hoping for a 0.016(6) reload. It appeared to shoot once per frame, which should be the case, but the box was emptied in 16.66(6) seconds, or 1000 frames. The theoretical shoting time should've been 8.33(3) seconds or 500 frames. So 1 shot per 2 60 Hz frames.

Note: I've accounted for the fact that a second has 60 frames.

**Comment by Regalis11** [comment]
>Coilgun has a 0.25 s reload, not 0.15 s. 

Ah, you're right, I got the values in your post mixed up.

>Actually, 0.25 (coilgun) is dividable by 1/60 and so is 0.1 (chaingun).

In theory they are, but in practice not, because floating point values have finite precision.

But I would still say that the guns are working as expected: we haven't tuned the reload values to meet some theoretical rate of fire value, but to get the actual in-game firing rate to a level we feel is suitable. If you for some reason want a gun to have a very specific firing rate, you need to take the effects of Euler integration, floating point precision, the game's update rate and in which order the calculations and checks on the reload timer are done into account.

**Comment by Krzeszny** [comment]
Ok, this issue can be:
a) closed as "won't fix", or rather "can't fix"
b) be left open indefinitely for informational purposes (in case someone else notices the chaingun doesn't fire 10 shots per second) or until tick rate gets increased in the far future (probably not)
EDIT: so you chose c). That's fine with me.
